crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2384 @@
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    noncelen = hpke_aead_tab[aead_ind].Nn;
+    erv = ossl_hpke_labeled_expand(kctx,
+                                   nonce, noncelen,
+                                   secret, secretlen, /* salt */
+                                   OSSL_HPKE_SEC51LABEL, /* protocol label */
+                                   suitebuf, 6, /* suiteid */
+                                   OSSL_HPKE_NONCE_LABEL, /* label */
+                                   ks_context, ks_contextlen); /* ikmlen */
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    seqlen = hpke_seq2buf(ctx->seq, seqbuf, 12);
+    if (seqlen == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    } else {
+        size_t sind;
+        unsigned char cv;
+
+        if (seqlen > noncelen) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        /* non constant time - does it matter? maybe no */
+        for (sind = 0; sind != noncelen; sind++) {
+            if (sind < seqlen) {
+                cv = seqbuf[seqlen - 1 - (sind % seqlen)];
+            } else {
+                cv = 0x00;
+            }
+            nonce[noncelen - 1 - sind] ^= cv;
+        }
+    }
+    keylen = hpke_aead_tab[aead_ind].Nk;
+    erv = ossl_hpke_labeled_expand(kctx,
+                                   key, keylen,
+                                   secret, secretlen, /* salt */
+                                   OSSL_HPKE_SEC51LABEL, /* protocol label */
+                                   suitebuf, 6, /* suiteid */
+                                   OSSL_HPKE_KEY_LABEL, /* label */
+                                   ks_context, ks_contextlen); /* ikmlen */
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (ctx->exportersec == NULL) {
+        ctx->exporterseclen = hpke_kdf_tab[kdf_ind].Nh;
+        ctx->exportersec = OPENSSL_malloc(ctx->exporterseclen);
+        if (ctx->exportersec == NULL) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        erv = ossl_hpke_labeled_expand(kctx,
+                                       ctx->exportersec, ctx->exporterseclen,
+                                       secret, secretlen,
+                                       OSSL_HPKE_SEC51LABEL,
+                                       suitebuf, 6,
+                                       OSSL_HPKE_EXP_LABEL,
+                                       ks_context, ks_contextlen);
+        if (erv != 1) {
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+    if (do_enc == 1) {
+        erv = hpke_aead_enc(ctx->libctx, ctx->propq, ctx->suite,
```

Comment:
To me - this last little bit here is the actual seal....
Everything before here could be done with a separate call (and then stored in the ctx)
If you had this API - your existing API could just use it, and then we would both be happy :)

Reply:
Is that in the context of the comment [here](https://github.com/openssl/openssl/pull/17172#discussion_r990846300)? 
If so, I think I disagree. It'd be wrong to require input of a peer public value when generating a sender's key pair. It also seems wrong to me to require generation of a new key pair for every first seal operation to a given recipient public value; noting we don't quite have that now thanks to the IKM mechanism - but that's arguably putting the random number generation at the wrong layer of the s/w stack (at least for ECH where the AAD input to the 1st seal operation includes the sender's public value).

This could be a thing less well suited to comment based systems like this though. (email or voice may be better)